# Assignments
As in previous checkpoints, please submit links to two Juypyter notebooks (one for each assignment below).

Please submit links to all your work below. This is not a graded checkpoint, but you should discuss your solutions with your mentor. Also, when you're done, compare your work to these example solutions.

### 1. Weather model
For this assignment, you'll revisit the historical temperature dataset. To complete this assignment, submit a link a Jupyter notebook containing your solutions to the following tasks:


- First, load the dataset from the weatherinszeged table from Thinkful's database.
- Like in the previous checkpoint, build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why?
- Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one?
- Add visibility as an additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the visibility in terms of the improvement in the adjusted R-squared. Which one is more useful?
- Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.


In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn import linear_model
import statsmodels.formula.api as smf
from sqlalchemy import create_engine
import seaborn as sns

# Display preferences.
%matplotlib inline
pd.options.display.float_format = '{:.2f}'.format

import warnings
warnings.filterwarnings(action="ignore")

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

df = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()

df.head(10)

date        summary preciptype  temperature  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy       rain         9.47   
1 2006-03-31 23:00:00+00:00  Partly Cloudy       rain         9.36   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy       rain         9.38   
3 2006-04-01 01:00:00+00:00  Partly Cloudy       rain         8.29   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy       rain         8.76   
5 2006-04-01 03:00:00+00:00  Partly Cloudy       rain         9.22   
6 2006-04-01 04:00:00+00:00  Partly Cloudy       rain         7.73   
7 2006-04-01 05:00:00+00:00  Partly Cloudy       rain         8.77   
8 2006-04-01 06:00:00+00:00  Partly Cloudy       rain        10.82   
9 2006-04-01 07:00:00+00:00  Partly Cloudy       rain        13.77   

   apparenttemperature  humidity  windspeed  windbearing  visibility  \
0                 7.39      0.89      14.12       251.00       15.83   
1                 7.23      0.86      14.26       259.00       15.83   
2                 9.38      0.89       3.93       204.00       14.96   
3                 5.94      0.83      14.10       269.00       15.83   
4                 6.98      0.83      11.04       259.00       15.83   
5                 7.11      0.85      13.96       258.00       14.96   
6                 5.52      0.95      12.36       259.00        9.98   
7                 6.53      0.89      14.15       260.00        9.98   
8                10.82      0.82      11.32       259.00        9.98   
9                13.77      0.72      12.53       279.00        9.98   

   loudcover  pressure                       dailysummary  
0       0.00   1015.13  Partly cloudy throughout the day.  
1       0.00   1015.63  Partly cloudy throughout the day.  
2       0.00   1015.94  Partly cloudy throughout the day.  
3       0.00   1016.41  Partly cloudy throughout the day.  
4       0.00   1016.51  Partly cloudy throughout the day.  
5       0.00   1016.66  Partly cloudy throughout the day.  
6       0.00   1016.72  Partly cloudy throughout the day.  
7       0.00   1016.84  Partly cloudy throughout the day.  
8       0.00   1017.37  Partly cloudy throughout the day.  
9       0.00   1017.22  Partly cloudy throughout the day.

- Like in the previous checkpoint, build a linear regression model where your target variable is the difference between the apparenttemperature and the temperature. As explanatory variables, use humidity and windspeed. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why?

ANS: R-squared: 0.288, adjusted R-squared: 0.288
because higher (adjusted) R-squared values are forvored but not too high. So the result is not very satisfactory.

In [3]:
# build the linear regression model
# use statsmodels
import statsmodels.api as sm

# We need to manually add a constant
# in statsmodels' sm

# Y is the target variable
Y = df['apparenttemperature'] - df['temperature']
# X is the feature set which includes
# is_male and is_smoker variables
X = df[['humidity','windspeed']]



X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

# plt.plot(Y)

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Mon, 03 Feb 2020   Prob (F-statistic):               0.00
Time:                        12:09:49   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.000       2.397       2.479
humidity      -3.0292      0.024   -126.479      0.000      -3.076      -2.982
windspeed     -0.1193      0.001   -176.164      0.000      -0.121      -0.118
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.267
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                          -0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Next, include the interaction of humidity and windspeed to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one?

ANS: yes, the model performance has improved. Since (adjusted) R-squared value has increased, while not increase to the extent that we would consider overfitting issue. Also, note that the AIC and BIC value has decreased, which is also a good sign. 

In [8]:
# build the linear regression model
# use statsmodels
# import statsmodels.api as sm

# We need to manually add a constant
# in statsmodels' sm

# Y is the target variable
Y = df['apparenttemperature'] - df['temperature']
# X is the feature set which includes
# is_male and is_smoker variables
X = pd.concat([df[['humidity','windspeed']], (df['humidity']*df['windspeed']).rename('interaction')], axis=1)



X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

# plt.plot(Y)

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Mon, 03 Feb 2020   Prob (F-statistic):               0.00
Time:                        12:20:14   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           0.0839      0.033      2.511      0.012       0.018       0.149
humidity        0.1775      0.043      4.133      0.000       0.093       0.262
windspeed       0.0905      0.002     36.797      0.000       0.086       0.095
interaction    -0.2971      0.003    -88.470      0.000      -0.304      -0.291
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.265
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                          -0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Add visibility as an additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the visibility in terms of the improvement in the adjusted R-squared. Which one is more useful?

ANS: yes, the R-squared value has increased from 0.288 to 0.304, the adjusted R-squared value has increased from 0.288 to 0.303. 
By comparing the result sumary after adding interaction varialbe vs. adding "visibility" varialbe, the interaction varialbe is more usefull, since the result has higher adjusted R-squared value, which is favored. So interaction variable is more useful.



In [10]:
# build the linear regression model
# use statsmodels
# import statsmodels.api as sm

# We need to manually add a constant
# in statsmodels' sm

# Y is the target variable
Y = df['apparenttemperature'] - df['temperature']
# X is the feature set which includes
# is_male and is_smoker variables
# X = pd.concat([df[['humidity','windspeed']], (df['humidity']*df['windspeed']).rename('interaction')], axis=1)
X = df[['humidity','windspeed', 'visibility']]


X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

# plt.plot(Y)

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Mon, 03 Feb 2020   Prob (F-statistic):               0.00
Time:                        12:24:45   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.000       1.521       1.630
humidity      -2.6066      0.025   -102.784      0.000      -2.656      -2.557
windspeed     -0.1199      0.001   -179.014      0.000      -0.121      -0.119
visibility     0.0540      0.001     46.614      0.000       0.052       0.056
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.282
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                          -0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

ANS: with respect to AIC and BIC, the smaller value is better. In that sense, model 2 > model 3 > model1, with AIC = 3.409 *10^5, 3.33*10^5, 3.38*10^5 respectively.

# 2. House prices model
In this exercise, you'll work on your house prices model. To complete this assignment, submit a link to a Jupyter notebook containing your solutions to the following tasks:

- Load the houseprices data from Thinkful's database.
- Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.
- Do you think your model is satisfactory? If so, why?
- In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.
For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

In [13]:
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'houseprices'

In [14]:
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
df = pd.read_sql_query('select * from houseprices',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()


# get rid of the columns with nan. (since we don't need that many features)
df_drop = df.copy()
for col in df_drop.columns:
    if df_drop[col].isna().any():
#         print(col)
        df_drop = df_drop.drop(columns=col)
    
    
# in this practice, let's just take two categorical variables, I picked "street", "lotshape"
# Encode character string categorical data into numeric data using one hot encoding
df_encoded = pd.concat([df_drop, pd.get_dummies(df_drop["street"], drop_first=True)], axis=1)
df_encoded = pd.concat([df_encoded, pd.get_dummies(df_encoded["lotshape"], drop_first=True)], axis=1)



In [15]:
# build the linear regression model
# use statsmodels

# Y is the target variable
Y = df_encoded['saleprice']
# X is the feature set which includes
# is_male and is_smoker variables
X = df_encoded[['Pave','IR2', 'IR3', 'Reg']]


X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.078
Model:                            OLS   Adj. R-squared:                  0.076
Method:                 Least Squares   F-statistic:                     30.90
Date:                Mon, 03 Feb 2020   Prob (F-statistic):           1.01e-24
Time:                        12:35:26   Log-Likelihood:                -18485.
No. Observations:                1460   AIC:                         3.698e+04
Df Residuals:                    1455   BIC:                         3.701e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.518e+05   3.14e+04      4.833      0.000    9.02e+04    2.13e+05
Pave        5.436e+04   3.13e+04      1.737      0.083   -7021.096    1.16e+05
IR2         3.495e+04   1.24e+04      2.809      0.005    1.05e+04    5.94e+04
IR3         9822.5103   2.44e+04      0.403      0.687    -3.8e+04    5.77e+04
Reg        -4.122e+04   4285.168     -9.620      0.000   -4.96e+04   -3.28e+04
==============================================================================
Omnibus:                      620.430   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3585.674
Skew:                           1.906   Prob(JB):                         0.00
Kurtosis:                       9.664   Cond. No.                         34.6
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### the R-squared value is way to low (close to 0), which indicates that the chosen features are not good to explain the target variable

### to fix this, let's include some other features, i.e. 'lotarea', 'yearbuilt'
the result showed that, our new model has higher adjusted R-squared value (0.338), and lower AIC, BIC values. So the new model is a better one. 

In [19]:
# build the linear regression model
# use statsmodels

# Y is the target variable
Y = df_drop['saleprice']
# X is the feature set which includes
# is_male and is_smoker variables
X = df_drop[['lotarea', 'yearbuilt']]


X = sm.add_constant(X)

results = sm.OLS(Y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              saleprice   R-squared:                       0.339
Model:                            OLS   Adj. R-squared:                  0.338
Method:                 Least Squares   F-statistic:                     373.9
Date:                Mon, 03 Feb 2020   Prob (F-statistic):          8.54e-132
Time:                        12:41:10   Log-Likelihood:                -18242.
No. Observations:                1460   AIC:                         3.649e+04
Df Residuals:                    1457   BIC:                         3.651e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const      -2.533e+06    1.1e+05    -22.935      0.000   -2.75e+06   -2.32e+06
lotarea        2.0412      0.170     12.041      0.000       1.709       2.374
yearbuilt   1365.7762     56.022     24.379      0.000    1255.883    1475.669
==============================================================================
Omnibus:                      741.084   Durbin-Watson:                   1.988
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             7175.267
Skew:                           2.155   Prob(JB):                         0.00
Kurtosis:                      12.969   Cond. No.                     9.51e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.51e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""